metadata/metadata.go

```diff
@@ -0,0 +1,145 @@
+type AggregationMetadata struct {
+	// List of aggregation types.
+	AggregationID aggregation.ID
+
+	// List of storage policies.
+	StoragePolicies []policy.StoragePolicy
+}
+
+// IsDefault returns whether this is the default aggregation metadata.
+func (m AggregationMetadata) IsDefault() bool {
+	return m.AggregationID.IsDefault() && policy.IsDefaultStoragePolicies(m.StoragePolicies)
+}
+
+// StandardPipelineMetadata contains standard pipeline metadata.
+type StandardPipelineMetadata struct {
+	AggregationMetadata
+	applied.Pipeline
+}
+
+// IsDefault returns whether this is the default standard pipeline metadata.
+func (sm StandardPipelineMetadata) IsDefault() bool {
+	return sm.AggregationMetadata.IsDefault() && sm.Pipeline.IsEmpty()
+}
+
+// ForwardingPipelineMetadata contains forwarding pipeline metadata.
+type ForwardingPipelineMetadata struct {
+	applied.Pipeline
+}
+
+// IsDefault returns whether this is the default forwarding pipeline metadata.
+func (fm ForwardingPipelineMetadata) IsDefault() bool {
+	return fm.Pipeline.IsEmpty()
+}
+
+// PipelineMetadataUnion is a union of different pipeline metadatas.
+type PipelineMetadataUnion struct {
+	Type       PipelineType
+	Standard   []StandardPipelineMetadata
+	Forwarding ForwardingPipelineMetadata
+}
+
+// IsDefault returns whether this is the default pipeline metadata.
+func (pu PipelineMetadataUnion) IsDefault() bool {
+	return (pu.Type == StandardType && len(pu.Standard) == 0) ||
+		(pu.Type == ForwardingType && pu.Forwarding.IsDefault())
+}
+
+// Metadata represents the metadata associated with a metric.
+type Metadata struct {
+	AggregationMetadata
+
+	Pipeline PipelineMetadataUnion
+}
+
+// IsDefault returns whether this is the default metadata.
+func (m Metadata) IsDefault() bool {
+	return m.AggregationMetadata.IsDefault() && m.Pipeline.IsDefault()
+}
+
+// ForwardMetadata represents the metadata information associated with forwarded metrics.
+type ForwardMetadata struct {
+	// List of aggregation types.
+	AggregationID aggregation.ID
+
+	// Storage policy.
+	StoragePolicy policy.StoragePolicy
+
+	// Pipeline of operations that may be applied to the metric.
+	Pipeline applied.Pipeline
+}
+
+// StagedMetadata represents metadata with a staged cutover time.
+type StagedMetadata struct {
+	Metadata
+
+	// Cutover is when the metadata is applicable.
+	CutoverNanos int64
+
+	// Tombstoned determines whether the associated metric has been tombstoned.
+	Tombstoned bool
+}
+
+// IsDefault returns whether this is a default staged metadata.
+func (sm StagedMetadata) IsDefault() bool {
+	return sm.CutoverNanos == 0 && !sm.Tombstoned && sm.Metadata.IsDefault()
+}
+
+// StagedMetadatas contains a list of staged metadatas.
+type StagedMetadatas []StagedMetadata
+
+// IsDefault determines whether the list of staged metadata is a default list.
+func (sms StagedMetadatas) IsDefault() bool {
+	return len(sms) == 1 && sms[0].IsDefault()
```

Comment:
should check len() ==0 as well?

Reply:
A default staged metadatas should always have at least one element. An empty list of staged metadata is an error and will result in the metric being discarded at the agg tier.